graph/src/components/store.rs

```diff
@@ -1448,6 +1451,19 @@ impl EntityCache {
         Ok(())
     }
 
+    /// A copy of the pending updates. Those are kept in a persistent data structure so this is a
+    /// cheap operation.
+    pub(crate) fn updates_snapshot(&self) -> im::HashMap<EntityKey, Option<Entity>> {
+        self.updates.clone()
+    }
+
+    pub(crate) fn restore_updates_snapshot(
```

Comment:
Good news is I implemented this without the persistent data structure, which really was bad for performance. Bad news is the code touched some tricky logic in the entity cache and is not super robust since it requires explicitly entering and exiting the handler. But I added and end-to-end test and plenty of asserts.